wfl/Database/MongoDBManager.cs

```diff
@@ -29,14 +29,22 @@ public MongoDBManager(string inputCollectionName, string inputConnectionString)
             return entity;
         }
 
-        public async Task<IEnumerable<T>> GetAllEntitiesAsync()
+        public async Task<IEnumerable<T>> GetEntitiesAsync(ICriteria<T> criteria)
         {
             IMongoCollection<T> collection = db.GetCollection<T>(collectionName);
-            IAsyncCursor<T> task = await collection.FindAsync(e => true, null);
+            IAsyncCursor<T> task;
+            if (criteria != null)
+            {
+               task = await collection.FindAsync(criteria.CreateFilter(), null);
+            }
+            else
+            {
+                task = await collection.FindAsync(e => true, null);
+            }
             IList<T> entities = task.ToList();
-            if(entities.Count > 0 && entities[0] is IContainerEntity)
+            if (entities.Count > 0 && entities[0] is IContainerEntity)
             {
-                foreach(T entity in entities)
+                foreach (T entity in entities)
```

Comment:
FYI, LINQ exposes an `OfType<T>` which will filter based on type. You can add it later though.